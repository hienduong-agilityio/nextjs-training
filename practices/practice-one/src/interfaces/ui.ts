export interface IInputGroupProps {
  placeholder?: string;
  value?: string;
  buttonText?: string;
  inputName?: string;
  isDisabled?: boolean;
  isLoading?: boolean;
  startIcon?: React.ReactNode;
  customClass?: {
    container?: string;
    inputContainer?: string;
    input?: string;
    button?: string;
  };
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void;
}
