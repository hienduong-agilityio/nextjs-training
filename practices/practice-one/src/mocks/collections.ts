import { Category } from '@/types/category';

type HotDeal = {
  name: Category | 'all';
  count: number;
};

export const HOT_DEALS: HotDeal[] = [
  { name: 'all', count: 2 },
  { name: 'beauty', count: 2 },
  { name: 'fragrances', count: 48 },
  { name: 'furniture', count: 95 },
  { name: 'groceries', count: 23 },
];

export const BRANDS = [
  { name: 'Nike', count: 99 },
  { name: 'Adidas', count: 99 },
  { name: 'Siemens', count: 99 },
];

export const FILTER_GROUP = [{ title: 'Hot Deals', items: HOT_DEALS }];
