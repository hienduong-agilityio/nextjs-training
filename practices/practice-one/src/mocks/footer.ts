import { ROUTE } from '@/constants';

export const FOOTER_NAVIGATION = [
  {
    title: 'Information',
    items: [
      { name: 'About Us', url: ROUTE.ABOUT_US },
      { name: 'Information', url: ROUTE.INFORMATION },
      { name: 'Privacy Policy', url: ROUTE.PRIVACY_POLICY },
      { name: 'Terms & Conditions', url: ROUTE.TERM_AND_CONDITIONS },
    ],
  },
  {
    title: 'Service',
    items: [
      { name: 'About Us', url: ROUTE.ABOUT_US },
      { name: 'Information', url: ROUTE.INFORMATION },
      { name: 'Privacy Policy', url: ROUTE.PRIVACY_POLICY },
      { name: 'Terms & Conditions', url: ROUTE.TERM_AND_CONDITIONS },
    ],
  },
  {
    title: 'My Account',
    items: [
      { name: 'About Us', url: ROUTE.ABOUT_US },
      { name: 'Information', url: ROUTE.INFORMATION },
      { name: 'Privacy Policy', url: ROUTE.PRIVACY_POLICY },
      { name: 'Terms & Conditions', url: ROUTE.TERM_AND_CONDITIONS },
    ],
  },
  {
    title: 'Our Offers',
    items: [
      { name: 'About Us', url: ROUTE.ABOUT_US },
      { name: 'Information', url: ROUTE.INFORMATION },
      { name: 'Privacy Policy', url: ROUTE.PRIVACY_POLICY },
      { name: 'Terms & Conditions', url: ROUTE.TERM_AND_CONDITIONS },
    ],
  },
];
