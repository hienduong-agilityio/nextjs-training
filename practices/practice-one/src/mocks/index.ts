export * from './product';
export * from './service';
export * from './footer';
export * from './collections';
