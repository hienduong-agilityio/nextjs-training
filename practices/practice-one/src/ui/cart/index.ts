export * from './CartTable';
export * from './OrderSummary';
export * from './VoucherInput';
export * from './EmptyCart';

export { default as CartItemRow } from './CartItemRow';
