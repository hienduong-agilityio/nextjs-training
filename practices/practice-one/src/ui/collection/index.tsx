export * from './LoadingCategoryGroup';
export * from './LoadingFilterSortBar';
