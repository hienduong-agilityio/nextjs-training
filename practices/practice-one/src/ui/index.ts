export * from './products/index';
export * from './collection/index';
