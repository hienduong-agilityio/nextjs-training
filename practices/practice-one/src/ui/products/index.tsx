export * from './CartAndFavoriteActions';
export * from './ProductImages';
export * from './ProductInfo';
export * from './QuantityControl';
export * from './SocialShare';
export * from './RelatedProduct';
export * from './ProductDetailTabs';
export * from './ProductReview';
export * from './ProductActions';
export * from './ProductLoading';
