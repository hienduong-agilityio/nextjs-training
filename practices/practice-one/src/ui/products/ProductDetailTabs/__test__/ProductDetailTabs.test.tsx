import { render, screen, fireEvent } from '@testing-library/react';
import { useSearchParams } from 'next/navigation';

// UI
import { ProductDetailTabs } from '@/ui';

// Mocks
import { PRODUCT_REVIEW } from '@/mocks';

// Mock useSearchParams from Next.js
jest.mock('next/navigation', () => ({
  useSearchParams: jest.fn(),
}));

describe('ProductDetailTabs Component', () => {
  const mockDescription = 'This is a product description.';

  const mockSearchParams = (query: string) => {
    (useSearchParams as jest.Mock).mockReturnValue(new URLSearchParams(query));
  };

  const renderProductDetailTabs = (query: string = '') => {
    mockSearchParams(query);
    return render(
      <ProductDetailTabs
        description={mockDescription}
        reviews={PRODUCT_REVIEW}
      />,
    );
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('matches the snapshot', () => {
    const { asFragment } = renderProductDetailTabs();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the component correctly with default active tab', () => {
    renderProductDetailTabs();

    const tabsWrapper = screen.getByRole('tablist');
    expect(tabsWrapper).toHaveClass('flex flex-col w-full');

    const activeTab = screen.getByText('Product Information');
    expect(activeTab).toBeInTheDocument();
    expect(screen.getByText(mockDescription)).toBeInTheDocument();
    expect(activeTab).toHaveClass('text-primary-400 border-primary-400');

    const reviewsTab = screen.getByText('Reviews');
    fireEvent.click(reviewsTab);

    PRODUCT_REVIEW.forEach((review) => {
      expect(screen.getByText(review.reviewerName)).toBeInTheDocument();
    });
  });

  it('renders the reviews tab when the active tab is set to Reviews', () => {
    renderProductDetailTabs('?activeTab=Reviews');

    expect(screen.getByText('Reviews')).toBeInTheDocument();

    PRODUCT_REVIEW.forEach((review) => {
      expect(screen.getByText(review.reviewerName)).toBeInTheDocument();
    });
  });
});
