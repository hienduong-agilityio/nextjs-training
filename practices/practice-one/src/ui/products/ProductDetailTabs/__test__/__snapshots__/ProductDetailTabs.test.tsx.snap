// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetailTabs Component matches the snapshot 1`] = `
<DocumentFragment>
  <section
    class="mt-8 px-4 rounded-md py-8 bg-secondary-50"
  >
    <div
      class="flex flex-col w-full"
      role="tablist"
    >
      <ul
        class="flex justify-start gap-7 overflow-x-auto overflow-y-hidden whitespace-nowrap scrollbar-hide w-auto p-3"
      >
        <li>
          <a
            aria-controls="tabpanel-0"
            aria-selected="true"
            class="p-3 text-sm font-medium transition-colors duration-200 text-center outline-none border-b-2 capitalize hover:text-primary-100 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-black text-primary-400 border-primary-400"
            href="?activeTab=Product Information"
            id="tab-0"
            role="tab"
            tabindex="0"
          >
            Product Information
          </a>
        </li>
        <li>
          <a
            aria-controls="tabpanel-1"
            aria-selected="false"
            class="p-3 text-sm font-medium transition-colors duration-200 text-center outline-none capitalize hover:text-primary-100 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-black border-b-0"
            href="?activeTab=Reviews"
            id="tab-1"
            role="tab"
            tabindex="0"
          >
            Reviews
          </a>
        </li>
      </ul>
      <div
        class="text-secondary-500 px-3"
      >
        <div
          class="text-sm leading-6 pt-7 border-t-2"
          role="tabpanel"
        >
          <p>
            This is a product description.
          </p>
        </div>
        <div
          class="hidden"
          hidden=""
          role="tabpanel"
        >
          <div
            class="space-y-6"
          >
            <div
              class="border-b border-gray-200 pb-6 last:border-none "
            >
              <div
                class="flex flex-col sm:flex-row sm:items-center justify-between"
              >
                <h4
                  class="text-lg font-semibold text-gray-800"
                >
                  John Doe
                </h4>
                <span
                  class="text-sm text-gray-500 mt-1 sm:mt-0"
                >
                  1/1/2024
                </span>
              </div>
              <p
                class="text-sm text-gray-500 mt-1"
              >
                johndoe@example.com
              </p>
              <div
                class="mt-3 flex items-center"
              >
                <div
                  class="flex gap-2 space-x-2"
                >
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                </div>
              </div>
              <p
                class="mt-4 text-gray-700 text-base leading-relaxed"
              >
                Amazing product! Highly recommend.
              </p>
            </div>
            <div
              class="border-b border-gray-200 pb-6 last:border-none "
            >
              <div
                class="flex flex-col sm:flex-row sm:items-center justify-between"
              >
                <h4
                  class="text-lg font-semibold text-gray-800"
                >
                  Alice Johnson
                </h4>
                <span
                  class="text-sm text-gray-500 mt-1 sm:mt-0"
                >
                  1/2/2024
                </span>
              </div>
              <p
                class="text-sm text-gray-500 mt-1"
              >
                alicej@example.com
              </p>
              <div
                class="mt-3 flex items-center"
              >
                <div
                  class="flex gap-2 space-x-2"
                >
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                </div>
              </div>
              <p
                class="mt-4 text-gray-700 text-base leading-relaxed"
              >
                Absolutely fantastic product!
              </p>
            </div>
            <div
              class="border-b border-gray-200 pb-6 last:border-none "
            >
              <div
                class="flex flex-col sm:flex-row sm:items-center justify-between"
              >
                <h4
                  class="text-lg font-semibold text-gray-800"
                >
                  Bob Smith
                </h4>
                <span
                  class="text-sm text-gray-500 mt-1 sm:mt-0"
                >
                  1/3/2024
                </span>
              </div>
              <p
                class="text-sm text-gray-500 mt-1"
              >
                bobsmith@example.com
              </p>
              <div
                class="mt-3 flex items-center"
              >
                <div
                  class="flex gap-2 space-x-2"
                >
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#C1C8CE"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                </div>
              </div>
              <p
                class="mt-4 text-gray-700 text-base leading-relaxed"
              >
                Great product, but the shipping was slow.
              </p>
            </div>
            <div
              class="border-b border-gray-200 pb-6 last:border-none "
            >
              <div
                class="flex flex-col sm:flex-row sm:items-center justify-between"
              >
                <h4
                  class="text-lg font-semibold text-gray-800"
                >
                  Charlie Brown
                </h4>
                <span
                  class="text-sm text-gray-500 mt-1 sm:mt-0"
                >
                  1/4/2024
                </span>
              </div>
              <p
                class="text-sm text-gray-500 mt-1"
              >
                charlieb@example.com
              </p>
              <div
                class="mt-3 flex items-center"
              >
                <div
                  class="flex gap-2 space-x-2"
                >
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#FFC600"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                  <svg
                    fill="none"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M41.982 17.527 29.044 27.992 32.85 45 20.28 34.725 6.9 44.453l5.169-16.817L0 16.642l15.764-.119L21.688 0l4.574 16.746 15.72.782Z"
                      fill="#C1C8CE"
                      transform="scale(0.36363636363636365)"
                    />
                  </svg>
                </div>
              </div>
              <p
                class="mt-4 text-gray-700 text-base leading-relaxed"
              >
                Good quality, but not worth the price.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
